File: Sources/XDecodeCore/DecodeLimits.swift
import Foundation

enum DecodeLimits {
    static let maximumInputFileSize: UInt64 = 500 * 1024 * 1024
    static let maximumDecompressedOutputSize = 1024 * 1024 * 1024

    static func validateInputFile(at url: URL, description: String) throws {
        let values: URLResourceValues
        do {
            values = try url.resourceValues(forKeys: [.fileSizeKey])
        } catch {
            throw DecodeError.fileOperation("无法读取\(description)大小：\(error.localizedDescription)")
        }
        guard let fileSize = values.fileSize, fileSize >= 0 else {
            throw DecodeError.fileOperation("无法读取\(description)大小")
        }
        guard UInt64(fileSize) <= maximumInputFileSize else {
            throw DecodeError.decodingFailed("\(description)超过 500 MB 大小限制")
        }
    }

    static func validateDecompressedOutputSize(
        currentSize: Int,
        appending additionalSize: Int,
        maximumSize: Int = maximumDecompressedOutputSize
    ) throws {
        let (combinedSize, overflow) = currentSize.addingReportingOverflow(additionalSize)
        guard currentSize >= 0,
              additionalSize >= 0,
              maximumSize >= 0,
              !overflow,
              combinedSize <= maximumSize else {
            throw DecodeError.outputLimitExceeded
        }
    }
}

File: Windows/src/XDecode.Core/DecodeLimits.cs
namespace XDecode.Core;

public static class DecodeLimits
{
    public const long MaximumInputFileSize = 500L * 1024 * 1024;
    public const int MaximumDecompressedOutputSize = 1024 * 1024 * 1024;

    public static void ValidateInputFile(string path, string description)
    {
        FileInfo info;
        try { info = new FileInfo(path); }
        catch (Exception exception) { throw DecodeException.FileOperation($"无法读取{description}大小：{exception.Message}", exception); }
        if (!info.Exists) throw DecodeException.FileOperation($"无法读取{description}大小");
        if (info.Length > MaximumInputFileSize)
            throw DecodeException.DecodingFailed($"{description}超过 500 MB 大小限制");
    }

    public static void ValidateOutputSize(long currentSize, long additionalSize, long maximumSize = MaximumDecompressedOutputSize)
    {
        if (currentSize < 0 || additionalSize < 0 || maximumSize < 0 ||
            currentSize > maximumSize - additionalSize)
            throw DecodeException.OutputLimitExceeded();
    }
}
